### ISSUE TITLE
Statements provided by delegate should not close underlying connection

### ISSUE THREAD

**Issue by tomdcc** [jOOQ/jOOQ#7463]
### Expected behavior and actual behavior:

If attempting to execute some pre-statement code in an `ExecuteListener.start()`, subsequent queries will fail.

I tracked this down to the `ProviderEnabled*Statement` classes closing their underlying connection. This is not normal JDBC behaviour. The underlying connection should remain open when a statement is closed.

### Steps to reproduce the problem:

Use a listener something like this:

```java
class Foo extends DefaultExecuteListener {
    @Override
    public void start(ExecuteContext ctx) {
        try (Statement st = ctx.connection().createStatement()) {
            st.executeUpdate("SET SESSION some_setting TO 'some value'");
        }
    }
    @Override
    public void end(ExecuteContext ctx) {
        try (Statement st = ctx.connection().createStatement()) {
            st.executeUpdate("SET SESSION some_setting TO DEFAULT");
        }
    }
} 
```

### Versions:

- jOOQ: 3.10.1, code same on master AFAICS
- Java: 1.8
- Database (include vendor): Postgres, Linux PGDG rpms and Mac Postgres.app
- OS: Mac / Linux
- JDBC Driver : Postgres 42.1.4

**Comment by lukaseder** [comment]
Thanks for your report. Indeed, this has been a known issue for a while. The problem is that `ctx.connection()` returns the internally used `ProviderEnabled*` type, which is used internally to close the JDBC `ResultSet` and `Statement` after jOOQ is done with any particular query, as well as release the `Connection` (note that the connection isn't necessarily closed, it is released as in `ConnectionProvider.release()`)

jOOQ 3.11 will already fix this through #4277. E.g. the following test works with jOOQ 3.11:

```java
    public void testExecuteListenerRunningJDBCStatements() {
        assertEquals(1, create(new RunningJDBCStatements())
            .selectOne()
            .fetchOneInto(int.class));
    }

    public static class RunningJDBCStatements extends DefaultExecuteListener {

        @Override
        public void start(ExecuteContext ctx) {
            test(ctx);
        }

        @Override
        public void end(ExecuteContext ctx) {
            test(ctx);
        }

        private void test(ExecuteContext ctx) {
            try (Statement st = ctx.connection().createStatement();
                ResultSet rs = st.executeQuery(DSL.using(ctx.family()).selectOne().getSQL())) {

                assertTrue(rs.next());
                assertEquals(1, rs.getInt(1));
                assertFalse(rs.next());
            }
            catch (SQLException e) {
                throw new DataAccessException("SQLException", e);
            }
        }
    }
```

The workaround in earlier versions of jOOQ is to acquire the connection from your `ConnectionProvider` directly, or even from your `DataSource`

**Comment by lukaseder** [comment]
Fixed through #4277

**Comment by tomdcc** [comment]
Great! Thanks @lukaseder !
